src/ObjectPool/src/DefaultObjectPool.cs

```diff
@@ -71,29 +71,43 @@ private T GetViaScan()
         [MethodImpl(MethodImplOptions.NoInlining)]
         private T Create() => _fastPolicy?.Create() ?? _policy.Create();
 
-        public override void Return(T obj)
+        public override void Return(T obj) => ReturnCore(obj);
+
+        protected bool ReturnCore(T obj)
         {
             if (_isDefaultPolicy || (_fastPolicy?.Return(obj) ?? _policy.Return(obj)))
             {
-                if (_firstItem != null || Interlocked.CompareExchange(ref _firstItem, obj, null) != null)
+                if (_firstItem == null && Interlocked.CompareExchange(ref _firstItem, obj, null) == null)
+                {
+                    return true;        // returned to pool
+                }
+                else
                 {
-                    ReturnViaScan(obj);
+                    return ReturnViaScan(obj);
                 }
             }
+
+            return false;           // not returned to pool
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void ReturnViaScan(T obj)
+        private bool ReturnViaScan(T obj)
         {
             var items = _items;
-            for (var i = 0; i < items.Length && Interlocked.CompareExchange(ref items[i].Element, obj, null) != null; ++i)
+            for (var i = 0; i < items.Length; i++)
             {
+                if (Interlocked.CompareExchange(ref items[i].Element, obj, null) == null)
+                {
+                    return true;
+                }
```

Comment:
I assume the structure of this had to change because you need to return true now instead of exiting the loop? IDK from your earlier optimization work if putting this test and loop condition together was a significant improvement. 

You could also do something like

```
var returned = false;
for (var i = 0; i < (items.Length && !(returned = Interlocked.CompareExchange(ref items[i].Element, obj, null) != null)); i++)
{
}

return returned;
```

Yeah, it's "too clever" 😉  but it might be worthwhile.

Reply:
This might end up generating the same code anyway.

Reply:
Your cleverness produces a bit better code, but it still regresses by about 7%. So I'll move this part to the `DisposableObjectPool` and leave the `DefaultObjectPool` with the fast `Return`-code.

Reply:
From the JIT-dumps I saw that the "scan-methods" aren't inlined due `runtime dictionary lookup` (which comes from generics). When manually inlining them, perf rises a bit.